src/main/scala/org/constellation/infrastructure/endpoints/middlewares/PeerAuthMiddleware.scala

```diff
@@ -52,22 +54,58 @@ object PeerAuthMiddleware {
     }
 
   def responseSignerMiddleware[F[_]: Sync](
-    privateKey: PrivateKey
+    privateKey: PrivateKey,
+    sessionTokenService: SessionTokenService[F]
   )(http: HttpRoutes[F])(implicit C: ContextShift[F]): HttpRoutes[F] =
     Kleisli { req: Request[F] =>
-      http(req).flatMap { res =>
-        new Http4sResponseSigner[F](getGenerator(privateKey))
-          .sign(res)
+      val logger = Slf4jLogger.getLogger[F]
+
+      for {
+        res <- http(req)
+        oHeaderToken <- sessionTokenService.getOwnToken
+          .map(_.map(t => Header(XSessionToken.value, t.value)))
           .attemptT
           .toOption
+        newHeaders = oHeaderToken.fold(res.headers)(h => res.headers.put(h))
+        resWithHeader = res.copy(headers = newHeaders)
+        signedResponse <- new Http4sResponseSigner[F](getGenerator(privateKey), new ConstellationHttpCryptoConfig {})
+          .sign(resWithHeader)
+          .attemptT
+          .toOption
+      } yield signedResponse
+    }
+
+  def responseTokenVerifierMiddleware[F[_]](
+    client: Client[F],
+    sessionTokenService: SessionTokenService[F]
+  )(implicit F: Concurrent[F], C: ContextShift[F]): Client[F] =
+    Client { (req: Request[F]) =>
+      client.run(req).flatMap { response =>
+        Resource.liftF {
+          val peerIp = req.uri.authority.map(_.host.value).getOrElse("unknown")
+          val oHeaderToken = response.headers
+            .get(XSessionToken)
+            .map(h => new Token(h.value))
```

Comment:
I think you can use apply for case class like `.map(Token(_.value))`

Reply:
unfortunately you can't extract .value from a wildcard like that :(